metafetcher/fetcher.go

```diff
@@ -0,0 +1,136 @@
+// process a message.
+const maxAttempts = 3
+
+// Message is the format we use to tell the MetaFetcher to download
+// the metadata for a given skylink and then add its size to the used space of
+// a given user.
+type Message struct {
+	UserID    primitive.ObjectID
+	SkylinkID primitive.ObjectID
+	Attempts  uint8
+}
+
+// MetaFetcher is a background task that listens for messages on its queue and
+// then processes them.
+type MetaFetcher struct {
+	Queue chan Message
+
+	db     *database.DB
+	portal string
+	logger *logrus.Logger
+}
+
+// New returns a new MetaFetcher instance and starts its internal queue watcher.
+func New(ctx context.Context, db *database.DB, portal string, logger *logrus.Logger) *MetaFetcher {
+	if logger == nil {
+		logger = logrus.New()
+	}
+	mf := MetaFetcher{
+		Queue:  make(chan Message, 1000),
+		db:     db,
+		portal: portal,
+		logger: logger,
+	}
+
+	go mf.threadedStartQueueWatcher(ctx)
+
+	return &mf
+}
+
+// threadedStartQueueWatcher starts a loop over the Queue that processes each
+// incoming message in a separate goroutine.
+func (mf *MetaFetcher) threadedStartQueueWatcher(ctx context.Context) {
+	for m := range mf.Queue {
+		// Process each message in a separate goroutine because fetching the
+		// meta might take a long time (30 seconds) and we don't want to block
+		// the queue.
+		go mf.processMessage(ctx, m)
+	}
+}
+
+// processMessage tries to download the metadata for the given skylink and
+// update the skylink's record in the database. It will also update the user's
+// used storage. If it fails to download it will put the message back in the
+// queue and retry it later a maximum of maxAttempts times.
+func (mf *MetaFetcher) processMessage(ctx context.Context, m Message) {
+	sl, err := mf.db.SkylinkByID(ctx, m.SkylinkID)
+	if err != nil {
+		logrus.Tracef("Failed to fetch skylink from DB. Skylink ID: %v, error: %v\n", m.SkylinkID, err)
+		if m.Attempts >= maxAttempts {
+			mf.logger.Debugf("Message exceeded its maximum number of attempts, dropping: %v\n", m)
+			return
+		}
+		m.Attempts++
+		mf.Queue <- m
+		return
+	}
+	// Make a HEAD request directly to the local `sia` container. We do that, so
+	// we don't get rate-limited by nginx in case we need to make many requests.
+	skylinkURL, err := url.Parse(fmt.Sprintf("http://sia:9980/skynet/skylink/%s", sl.Skylink))
+	if err != nil {
+		mf.logger.Debugf("Error while forming skylink URL for skylink %s. Error: %v\n", sl.Skylink, err)
+		return
+	}
+	req := http.Request{
+		Method: http.MethodHead,
+		URL:    skylinkURL,
+		Header: http.Header{"User-Agent": []string{"Sia-Agent"}},
+	}
+	client := http.Client{}
+	res, err := client.Do(&req)
+	if err != nil || res.StatusCode > 399 {
+		mf.logger.Tracef("Failed to fetch skyfile. Skylink: %s, error: %v\n", sl.Skylink, err)
+		if m.Attempts >= maxAttempts {
+			mf.logger.Debugf("Message exceeded its maximum number of attempts, dropping: %v\n", m)
```

Comment:
Should we consider `Errorf` here?

Reply:
`Error` is a bit severe for this. Maybe `Warn`? We intend to have a daily/weekly cover for this in the form of a scan that attempts to update the sizes on all zero-sized skylinks (#22).